File: swift/circular-buffer/Sources/CircularBuffer/CircularBuffer.swift
enum CircularBufferError: Error {
    case bufferEmpty, bufferFull
}

struct CircularBuffer<T> {
    var count = 0
    let capacity: Int
    var oldestIndex: Int
    var buffer: [T?]

    init(capacity: Int) {
        self.capacity = capacity
        self.oldestIndex = 0
        self.buffer = Array(repeating: nil, count: capacity)
    }
    
    mutating func read() throws -> T {
        if self.count == 0 {
            throw CircularBufferError.bufferEmpty
        }
        let res = self.buffer[self.oldestIndex]!
        self.oldestIndex = (self.oldestIndex + 1) % self.capacity
        self.count -= 1
        return res
    }
    
    mutating func write(_ item: T) throws {
        if (self.count == self.capacity) {
            throw CircularBufferError.bufferFull
        }
        self.buffer[(self.oldestIndex + self.count) % self.capacity] = item
        self.count += 1
    }
    
    mutating func overwrite(_ item: T) {
        if (self.count < self.capacity) {
            try? self.write(item)
        } else {
            self.buffer[self.oldestIndex] = item
            self.oldestIndex = (self.oldestIndex + 1) % self.capacity
        }
    }
    
    mutating func clear() {
        self.count = 0
        self.buffer = Array(repeating: nil, count: self.capacity)
        self.oldestIndex = 0
    }
}

File: typescript/circular-buffer/circular-buffer.ts
export class BufferOverflowError extends Error {}
export class BufferEmptyError extends Error {}

export default class CircularBuffer<T> {
  buffer: T[]
  oldestIndex: number
  count: number = 0

  constructor(private readonly capacity: number) {
    this.buffer = new Array(capacity)
    this.oldestIndex = Math.floor(Math.random() * this.capacity)
  }
  
  read(): T {
    if (!this.count) {
      throw new BufferEmptyError()
    }
    const res = this.buffer[this.oldestIndex]
    this.count--
    this.oldestIndex = (this.oldestIndex + 1) % this.capacity
    return res
  }

  write(element: T): void {
    if (this.isFull()) {
      throw new BufferOverflowError()
    }
    this.buffer[(this.oldestIndex + this.count) % this.capacity] = element
    this.count++
  }

  forceWrite(element: T): void {
    if (!this.isFull()) {
      this.write(element)
    } else {
      this.buffer[this.oldestIndex] = element
      this.oldestIndex = (this.oldestIndex + 1) % this.capacity
    }
  }

  clear() {
    this.count = 0;
  }

  isFull() {
    return this.count === this.capacity
  }
}